components/dashboard/src/workspaces/StartWorkspaceModal.tsx

```diff
@@ -4,31 +4,93 @@
+import { useContext, useEffect, useMemo, useState } from "react";
 import { useLocation } from "react-router";
 import Modal from "../components/Modal";
 import RepositoryFinder from "../components/RepositoryFinder";
+import SelectIDEComponent from "../components/SelectIDEComponent";
+import SelectWorkspaceClassComponent from "../components/SelectWorkspaceClassComponent";
+import { UserContext } from "../user-context";
 import { StartWorkspaceModalContext } from "./start-workspace-modal-context";
 
 export function StartWorkspaceModal() {
+    const { user } = useContext(UserContext);
     const { isStartWorkspaceModalVisible, setIsStartWorkspaceModalVisible } = useContext(StartWorkspaceModalContext);
     const location = useLocation();
+    const [useLatestIde, setUseLatestIde] = useState<boolean | undefined>(
+        !!user?.additionalData?.ideSettings?.useLatestVersion,
+    );
+    const [selectedIde, setSelectedIde] = useState(user?.additionalData?.ideSettings?.defaultIde);
+    const [selectedWsClass, setSelectedWsClass] = useState<string | undefined>();
+    const [repo, setRepo] = useState("");
+    const onSelectEditorChange = (ide: string, useLatest: boolean) => {
+        setSelectedIde(ide);
+        setUseLatestIde(useLatest);
+    };
+
+    const startWorkspace = useMemo(() => {
+        return () => {
+            if (!repo) {
+                return false;
+            }
+            const url = new URL(window.location.href);
+            url.pathname = "";
+            const searchParams = new URLSearchParams();
+            if (selectedWsClass) {
+                searchParams.set(StartOptions.WORKSPACE_CLASS, selectedWsClass);
+            }
+            if (selectedIde) {
+                searchParams.set(StartOptions.EDITOR, selectedIde);
+                searchParams.set(StartOptions.USE_LATEST_EDITOR, useLatestIde ? "true" : "false");
+            }
+            url.search = searchParams.toString();
+            url.hash = "#" + repo;
+            window.location.href = url.toString();
+            return true;
+        };
+    }, [repo, selectedIde, selectedWsClass, useLatestIde]);
 
     // Close the modal on navigation events.
     useEffect(() => {
         setIsStartWorkspaceModalVisible(false);
-    }, [location]);
+    }, [location, setIsStartWorkspaceModalVisible]);
 
     return (
-        // TODO: Use title and buttons props
         <Modal
             onClose={() => setIsStartWorkspaceModalVisible(false)}
-            onEnter={() => false}
+            onEnter={startWorkspace}
             visible={!!isStartWorkspaceModalVisible}
+            title="Open in Gitpod"
+            buttons={[
+                <button key="cancel" className="secondary" onClick={() => setIsStartWorkspaceModalVisible(false)}>
+                    Cancel
+                </button>,
+                <button key="start" className="" onClick={startWorkspace}>
+                    New Workspace
```

Comment:
**suggestion**: Could we disable the New Workspace button when there's no repository selected? Currently, having this always enabled can lead to broken UX.

Reply:
Yes